src/compiler/checker.ts

```diff
@@ -11225,6 +11234,56 @@ namespace ts {
             }
         }
 
+        /**
+         * Determines if all declarations in the given list are function declarations, and that
+         * all the declarations have the same name. This is useful for how we report errors.
+         *
+         * For instance, if a user has
+         *
+         *      export default function foo() { }
+         *      export default function bar() { }
+         *
+         * We want to tell the user that there are multiple default exports. However, if a user has
+         *
+         *      export default function foo() { }
+         *      export default function foo() { }
+         *
+         * It is probably more accurate to say that there are duplicate function implementations.
+         * 
+         * @param declarations
+         */
+        function areAllFunctionDeclarationsWhereSomeNamesDiffer(declarations: Declaration[]) {
+            let numDeclarations = declarations.length;
+            Debug.assert(numDeclarations > 0, "Expected at least one declaration.");
+
+            let firstDeclaration = declarations[0];
+            if (firstDeclaration.kind !== SyntaxKind.FunctionDeclaration) {
+                return false;
+            }
+
+            let firstName = (firstDeclaration as FunctionDeclaration).name;
+            for (let i = 1; i < numDeclarations; i++) {
+                let currentDeclaration = declarations[i];
+                if (currentDeclaration.kind !== SyntaxKind.FunctionDeclaration) {
+                    return false;
+                }
+
+                let currentName = (currentDeclaration as FunctionDeclaration).name;
+                if (currentName) {
+                    if (!firstName || currentName.text !== firstName.text) {
+                        // Either 'firstName' must be undefined or the text must differ.
+                        return true;
+                    }
+                }
+                else if (firstName) {
+                    // 'currentName' is not defined but 'firstName' is.
+                    return true;
+                }
+            }
+
+            return false;
+        }
+

```

Comment:
Checking both properties in this loop is hard to read. Why not use `every` and `some`, since the function name already expresses those concepts? I propose this rewrite:

        function areAllFunctionDeclarationsWhereSomeNamesDiffer(declarations: Declaration[]) {
            Debug.assert(declarations.length > 0, "Expected at least one declaration.");
            let firstName = (declarations[0] as FunctionDeclaration).name;
            return declarations.every(d => d.kind === SyntaxKind.FunctionDeclaration) &&
                declarations.some(d => {
                    let currentName = (d as FunctionDeclaration).name;
                    if (currentName && firstName) {
                        // Either 'firstName' must be undefined or the text must differ.
                        return currentName.text !== firstName.text;
                    }
                    else if (currentName || firstName) {
                        // one undefined, the other defined
                        return true;
                    }
                    else {
                        // both undefined
                        return false;
                    }
                });
        }


This also makes the conditionals easier to read by avoiding nesting (mostly).